src/checkpoint/dispatch/dispatch_virtual.h

```diff
@@ -48,4 +48,115 @@
+ *
+ *     Example with raw pointer:
+ *
+ *       template <typename T>
+ *       struct MyObjectWithRawPointer {
+ *         T* raw_ptr;
+ *
+ *         template <typename SerializerT>
+ *         void serialize(SerializerT& s) {
+ *           bool is_null = raw_ptr == nullptr;
+ *           s | is_null; // serialize whether we have a null pointer not
+ *           if (!is_null) {
+ *             // During size/pack, save the actual derived type of raw_ptr;
+ *             // During unpack, allocate/construct raw_ptr with correct virtual type
+ *             checkpoint::allocateConstructForPointer(s, raw_ptr);
+ *             s | *raw_ptr;
+ *           }
+ *         }
+ *       };
+ *
+ * ---------------------------------------------------------------------------
+ * --------------------- Overview of implementation: -------------------------
+ * ---------------------------------------------------------------------------
+ *
+ * When packing we need to traverse the whole hierarchy of virtual classes of
+ * serialize all the user's content. Thus, \c SerializableBase<T> and
+ * \c SerializableDervived<T,BaseT> insert a shim layer between all layers of
+ * the virtual hierarchy and below the base class. The alternative macros,
+ * instead of inserting a shim layer, just add the required virtual methods for
+ * recursively serializing the hierarchy.
+ *
+ * The macros or shim classes insert a virtual method
+ * \c _checkpointDynamicSerialize which is called during sizing, packing, and
+ * unpacking (all traversals). \c _checkpointDynamicSerialize calls the
+ * serializer on the base class recursively to get the whole hierarchy
+ * serialized.
+ *
+ * To actually dispatch the to the object's serialize method, there is a problem
+ * that the serializer type is templated (and is non-virtual). Since we can't
+ * mix virtual and templates, the serializer is carried through
+ * \c _checkpointDynamicSerialize as a \c void* but then dispatched using a
+ * type registry for the serializers. The type registry, templated on the object
+ * type, type-erases the templated serializer type (see serializer_registry.h).
+ * To ensure that all serializers are instantiated properly,
+ * \c _checkpointDynamicSerialize calls
+ * \c checkpoint::instantiateObjSerializer<T,checkpoint_serializer_variadic_args()>
+ * where \c checkpoint_serializer_variadic_args() is a macro that contains a
+ * pack of all existing serializers that ship with checkpoint.
+ *
+ * *Important*: if you define a custom traversal, you need to manually
+ * instantiate it for all virtual objects.
+ *
+ * When a serializer is looked up in the registry, it gets an index for a given
+ * \c T -- where \c T is a user's virtually serialized object type. However, we
+ * save the registry index for the base class in the hierarchy. Because the
+ * serializer registry is templated on \c T (i.e., there is a separate registry
+ * for each \c T ) we must match up the entry across registries for a derived
```

Comment:
Maybe I'm being dense or forgetful, but why can't we have a typedef in `SerializableBase<BaseT>`, `using _checkpointUltimateBaseType = BaseT`, and then always refer to registries directly by that ultimate base type?

I feel like something like that was there, and I can't tell if that was taken out because it doesn't work for some reason, or just because it got dropped on the floor.

Reply:
If I understand your remark correctly---I started with that, which in theory would have simplified things.

However, it creates undefined behavior because you have a dependency between two static initializations (the ordering issue). Basically, while you statically initializing `DerviedT`, you try to generated the index for the base---static initialize `BaseT`.

Reply:
Oh, if you're saying we should never register for the derived at all and just use the base, that's not possible because the operator| with SerializerT must be invoked on the *derived* type with that exact type.

Reply:
I agree, we clearly need to register every derived type, to actually invoke things properly.

I'm unclear about the UB that would arise from consolidating the registry. Do you mean true UB, or just that the order and hence indices would be unspecified? I think we might be able to impose some sort of natural ordering, such as the ultimate base always being index 0, and a topological ordering in which parents in turn precede children.

Reply:
https://isocpp.org/wiki/faq/ctors#static-init-order makes me think that this should be doable. If there's some subtlety that I'm missing, it sounded like you maybe had examples in hand?

Reply:
The registry is already in a function:
```C++
template <typename ObjT>
inline RegistryType<ObjT>& getRegistry() {
  static RegistryType<ObjT> reg;
  return reg;
}
```

I'm not sure how to phase the issue exactly (maybe it's a dynamic initialization dependency), but getting the index from another registry during `Registrar<ObjT, SerializerT>::Registrar()`, which happens at load time is definitely causing UB, because I see it generating zeros sometimes or the correct value other times which is the typical outcome from doing something like this.

Reply:
I'm willing to accept this for this PR, but I'd let to get a better understanding of what's going on with you sometime soon.

Reply:
I can write up a tiny example that shows the problem.